### PR TITLE
Streaming rich function

### PR DESCRIPTION
**Pull request by chaokunyang**
## Why are these changes needed?
Add `RichFunction` to  define the life cycle methods of user function

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by chaokunyang** [comment]
@BalaBalaYi  Could you help review this PR?

File: streaming/python/runtime/task.py [hunk 1]
```diff
@@ -83,9 +83,11 @@ def exit_handler():
             import atexit
             atexit.register(exit_handler)
 
+        # TODO(chaokunyang) add task/job config
         runtime_context = RuntimeContextImpl(
             self.worker.execution_task.task_id,
-            self.worker.execution_task.task_index, execution_node.parallelism)
+            self.worker.execution_task.task_index, execution_node.parallelism,
```
Comment (BalaBalaYi) [inline streaming/python/runtime/task.py:89]:
```
a,b,c,d 
```
or
```
a,
b,
c,
d
```
for better readable?

Comment (chaokunyang) [reply streaming/python/runtime/task.py]: Since `a,b,c,d` here will be exceed line width `79`, so I put it in next line

Comment (chaokunyang) [reply streaming/python/runtime/task.py]: It's also enforced by python lint

File: streaming/java/streaming-api/src/main/java/io/ray/streaming/api/function/RichFunction.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * functions, and access the task context where the functions get executed.
+ */
+public interface RichFunction extends Function {
+
+  /**
+   * Initialization method for user function which called before the first call to the user
+   * function.
+   * @param runtimeContext runtime context
+   */
+  void open(RuntimeContext runtimeContext);
```
Comment (BalaBalaYi) [inline streaming/java/streaming-api/src/main/java/io/ray/streaming/api/function/RichFunction.java:16]: Should we add a 'init' method for initialization with a context or just merge 'init' into 'open'?

File: streaming/java/streaming-api/src/main/java/io/ray/streaming/api/function/internal/RichFunctions.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
+package io.ray.streaming.api.function.internal;
+
+import io.ray.streaming.api.context.RuntimeContext;
+import io.ray.streaming.api.function.Function;
+import io.ray.streaming.api.function.RichFunction;
+
+public class RichFunctions {
```
Comment (BalaBalaYi) [inline streaming/java/streaming-api/src/main/java/io/ray/streaming/api/function/internal/RichFunctions.java:7]: What is this class use for? Give some comments?

File: streaming/java/streaming-api/src/main/java/io/ray/streaming/api/function/RichFunction.java (hunk 2 above)
Comment (chaokunyang) [reply streaming/java/streaming-api/src/main/java/io/ray/streaming/api/function/RichFunction.java:16]: I think `init` can be put into `open`

File: streaming/java/streaming-api/src/main/java/io/ray/streaming/api/function/internal/RichFunctions.java (hunk 3 above)
Comment (chaokunyang) [reply streaming/java/streaming-api/src/main/java/io/ray/streaming/api/function/internal/RichFunctions.java:7]: It's a util class for `Function`, I think renaming it to `Functions` would be more readable.

**Review by BalaBalaYi — APPROVED** [review]

**Review by raulchen — APPROVED** [review]

**Pull request merged**